pkg/kubelet/container/interface.go

```diff
@@ -0,0 +1,290 @@
+// container.
+type ContainerInspectResult struct {
+	// ID of the container.
+	ID string
+	// Config of the container.
+	Config ContainerConfig
+	// Status of the container.
+	// TODO: Re-examine ContainerStatus and consolidate the fields.
+	Status ContainerStatus
+}
+
+type ContainerConfig struct {
+	// Name of the container.
+	Name string
+	// Image to use.
+	Image ImageSpec
+	// RootFSPath is the path to the root filesystem. This field is optional.
+	// If not set, the runtime should create the root filesystem itself.
+	RootFSPath *string
+	// Command to execute (i.e., entrypoint for docker)
+	Command []string
+	// Args for the Command (i.e., command for docker)
+	Args []string
+	// Current working directory of the command.
+	WorkingDir string
+	// List of environment variable to set in the container
+	Env []KeyValue
+	// Annotations is an unstructured key value map that may be set by external
+	// tools to store and retrieve arbitrary metadata.
+	Annotations map[string]string
+	// Mount specifies mounts for the container
+	Mounts []Mount
+	// Privileged runs the container in the privileged mode.
+	Privileged bool
+	// ReadOnlyRootFS sets the root filesystem of the container to be
+	// read-only.
+	ReadOnlyRootFS bool
+	// Linux contains configuration specific to Linux containers.
+	Linux *LinuxContainerConfig
+	// Path to store the container log on the host (i.e., outside of the
+	// sandbox).
+	LogPath string
+}
+
+// LinuxContainerConfig contains configurations speicifc to Linux containers.
+type LinuxContainerConfig struct {
+	// AddCapabilities lists capabilities to add.
+	AddCapabilities []string
+	// DropCapabilities lists capabilities to drop.
+	DropCapabilities []string
+}
+
+// ContainerFilter is used to filter containers.
+type ContainerFilter struct {
+	// Name of the container.
+	Name string
+	// ID of the container.
+	ID string
+	// State of the contianer.
+	State ContainerState
+	// Annotations of the container. A container needs to have all listed
+	// annotations to be considered a match.
+	Annotations map[string]string
+}
+
+type StreamOptions struct {
+	TTY          bool
+	InputStream  io.Reader
+	OutputStream io.Writer
+	ErrorStream  io.Writer
+}
+
+// KeyValue represents a key-value pair.
+type KeyValue struct {
+	Key   string
+	Value string
+}
+
+// ImageManager offers basic image operations.
+// TODO: Need a better name.
+type ImageManager interface {
+	// List lists the existing images.
+	List() ([]Image, error)
+	// Pull pulls an image with authentication config.
+	Pull(image ImageSpec, auth AuthConfig) error
+	// Remove removes an image.
+	Remove(image ImageSpec) error
+	// Inspect inspects an image.
+	Inspect(image ImageSpec) (Image, error)
+}
+
+// AuthConfig contains authorization information for connecting to a registry.
+// TODO: This is copied from docker's Authconfig. We should re-evaluate to
+// support other registries.
+type AuthConfig struct {
+	Username      string
+	Password      string
```

Comment:
Hmm...the runtime has to know what type we are passing anyway. I am not sure how using secret will help.